File: leetcode-solutions/ruby/232. Implement Queue using Stacks.rb
# https://leetcode.com/problems/implement-queue-using-stacks/

class MyQueue

  =begin
      Initialize your data structure here.
  =end
      def initialize()
          @p_stack = []
          @pop_stack = []
      end
  
  
  =begin
      Push element x to the back of queue.
      :type x: Integer
      :rtype: Void
  =end
      def push(x)
          @p_stack << x
      end
  
  
  =begin
      Removes the element from in front of queue and returns that element.
      :rtype: Integer
  =end
      def pop()
          if @pop_stack.empty?
              while !@p_stack.empty?
                  @pop_stack << @p_stack.pop
              end
          end
          
          @pop_stack.pop
      end
  
  
  =begin
      Get the front element.
      :rtype: Integer
  =end
      def peek()
          if @pop_stack.empty?
              while !@p_stack.empty?
                  @pop_stack << @p_stack.pop
              end
          end
          
          @pop_stack.last
      end
  
  
  =begin
      Returns whether the queue is empty.
      :rtype: Boolean
  =end
      def empty()
          @pop_stack.empty? && @p_stack.empty?
      end
  
  
  end
  
  # Your MyQueue object will be instantiated and called as such:
  # obj = MyQueue.new()
  # obj.push(x)
  # param_2 = obj.pop()
  # param_3 = obj.peek()
  # param_4 = obj.empty()

File: leetcode-solutions/python/232. Implement Queue using Stacks.py
# https://leetcode.com/problems/implement-queue-using-stacks/

class MyQueue:

    def __init__(self):
        self.enqueueStack = []
        self.dequeueStack = []

    def push(self, x: int) -> None:
        self.enqueueStack.append(x)

    def pop(self) -> int:
        if not self.dequeueStack:
            self.transferEntoDeQueue()

        if not self.dequeueStack:
            return -1
        return self.dequeueStack.pop()
            

    def peek(self) -> int:
        if not self.dequeueStack:
            self.transferEntoDeQueue()

        return self.dequeueStack[-1]

    def empty(self) -> bool:
        return not(self.enqueueStack or self.dequeueStack)
    
    def transferEntoDeQueue(self):
        while self.enqueueStack:
            self.dequeueStack.append(self.enqueueStack.pop())

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()